File: Python/XGBoost/XGBoost.py
import pandas as pd
from sklearn.metrics import mean_squared_error
import xgboost as xgb
import matplotlib.pyplot as plt

# Load the dataset
file_path = "../../Dataset/XGBoost/airPassengers.csv"
data = pd.read_csv(file_path)

# Convert the "month" column to datetime format
data["month"] = pd.to_datetime(data["month"])

# Set "month" as the time index
data.set_index("month", inplace=True)

# Plot the time series data from the dataset
plt.figure(figsize=(10, 6))
plt.plot(data, label="Air Passengers")
plt.title("Air Passengers Time Series")
plt.xlabel("Year")
plt.ylabel("Number of Passengers")
plt.legend()
plt.show()

# Feature engineering: add time-based features if needed
data["year"] = data.index.year
data["month"] = data.index.month
data["day"] = data.index.day

# Split the data into training and testing sets
train_size = int(len(data) * 0.8)
train, test = data[0:train_size], data[train_size:]

# Define features and target variable
features = ["year", "month", "day"]
target = "passengers"

# Defining the training and test data
X_train, y_train = train[features], train[target]
X_test, y_test = test[features], test[target]

# Creating and train the XGBoost model
model = xgb.XGBRegressor(objective="reg:squarederror")
model.fit(X_train, y_train)

# Making predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error on Test Set: {mse}")

# Plot actual vs. predicted values
plt.figure(figsize=(10, 6))
plt.plot(test.index, y_test, label="Actual")
plt.plot(test.index, y_pred, label="Predicted")
plt.title("XGBoost Time Series Prediction")
plt.xlabel("Year")
plt.ylabel("Number of Passengers")
plt.legend()
plt.show()

File: Julia/XGBoost/XGBoost.jl
using XGBoost, DataFrames, CSV, Dates, Plots, MLLabelUtils

# Loading the Dataset
df = CSV.File("Dataset/XGBoost/airPassengers.csv") |> DataFrame

# Converting 'month' column in Date type
df.month = Dates.Date.(string.(df.month), "yyyy-mm-dd")

# Conserver 'month' sous forme de dates
df.month = Dates.value.(df.month)

# Convert DataFrame to matrices 
X = select(df, Not(:passengers))
y = df[!, :passengers]

# Define the model parameters
param = Dict(
    "objective" => "reg:squarederror",
    "max_depth" => 6,            # Tree maximum depth
    "eta" => 0.1,                # Learning rate
    "subsample" => 0.8,          # Fraction of samples used to train each tree
    "colsample_bytree" => 0.8    # Fraction of features used to train each tree
)

# Train the xgboost model
dtrain = DMatrix(X, label=y)
model = xgboost(dtrain, num_round=15, param=param)

# Making predictions with xgboost model
predictions = predict(model, dtrain)

println("Prédictions: ", predictions)

# Ploting the results
plot(df.month, y, label="Observations", xlabel="Date", ylabel="Passengers", legend=:top)
plot!(df.month, predictions, label="Prédictions", color=:red)
